[BUG] Exception when counting rows in an ORC file that has no column names

**Describe the bug**
Trying to count rows in an ORC file that does not have column names results in an ArrayIndexOutOfBoundsException like the following:
```
java.io.IOException: Error when processing file splits [path: file:///home/jlowe/src/spark-rapids/tests/src/test/resources/schema-cant-prune.orc, range: 0-323, partition values: [empty row]]
  at com.nvidia.spark.rapids.OrcCommonFunctions.decodeToTable(GpuOrcScan.scala:852)
  at com.nvidia.spark.rapids.OrcCommonFunctions.decodeToTable$(GpuOrcScan.scala:816)
  at com.nvidia.spark.rapids.MultiFileOrcPartitionReader.decodeToTable(GpuOrcScan.scala:1929)
  at com.nvidia.spark.rapids.MultiFileOrcPartitionReader.$anonfun$readBufferToTablesAndClose$1(GpuOrcScan.scala:2180)
  at com.nvidia.spark.rapids.Arm$.withResource(Arm.scala:29)
  at com.nvidia.spark.rapids.MultiFileOrcPartitionReader.readBufferToTablesAndClose(GpuOrcScan.scala:2178)
  at com.nvidia.spark.rapids.MultiFileCoalescingPartitionReaderBase.$anonfun$readBatch$4(GpuMultiFileReader.scala:1169)
  at com.nvidia.spark.rapids.RmmRapidsRetryIterator$AutoCloseableAttemptSpliterator.next(RmmRapidsRetryIterator.scala:424)
  at com.nvidia.spark.rapids.RmmRapidsRetryIterator$RmmRapidsRetryIterator.next(RmmRapidsRetryIterator.scala:528)
  at com.nvidia.spark.rapids.RmmRapidsRetryIterator$RmmRapidsRetryAutoCloseableIterator.next(RmmRapidsRetryIterator.scala:458)
  at com.nvidia.spark.rapids.RmmRapidsRetryIterator$.drainSingleWithVerification(RmmRapidsRetryIterator.scala:275)
  at com.nvidia.spark.rapids.RmmRapidsRetryIterator$.withRetryNoSplit(RmmRapidsRetryIterator.scala:128)
  at com.nvidia.spark.rapids.MultiFileCoalescingPartitionReaderBase.$anonfun$readBatch$1(GpuMultiFileReader.scala:1163)
  at com.nvidia.spark.rapids.Arm$.withResource(Arm.scala:29)
  at com.nvidia.spark.rapids.MultiFileCoalescingPartitionReaderBase.readBatch(GpuMultiFileReader.scala:1137)
  at com.nvidia.spark.rapids.MultiFileCoalescingPartitionReaderBase.next(GpuMultiFileReader.scala:1122)
  at com.nvidia.spark.rapids.PartitionIterator.hasNext(dataSourceUtil.scala:29)
  at com.nvidia.spark.rapids.MetricsBatchIterator.hasNext(dataSourceUtil.scala:46)
  at org.apache.spark.InterruptibleIterator.hasNext(InterruptibleIterator.scala:37)
  at org.apache.spark.sql.rapids.GpuFileSourceScanExec$$anon$1.hasNext(GpuFileSourceScanExec.scala:469)
  at com.nvidia.spark.rapids.GpuHashAggregateIterator.$anonfun$next$2(aggregate.scala:555)
  at scala.Option.getOrElse(Option.scala:189)
  at com.nvidia.spark.rapids.GpuHashAggregateIterator.next(aggregate.scala:553)
  at com.nvidia.spark.rapids.GpuHashAggregateIterator.next(aggregate.scala:498)
  at org.apache.spark.sql.rapids.execution.GpuShuffleExchangeExecBase$$anon$1.partNextBatch(GpuShuffleExchangeExecBase.scala:318)
  at org.apache.spark.sql.rapids.execution.GpuShuffleExchangeExecBase$$anon$1.hasNext(GpuShuffleExchangeExecBase.scala:340)
  at org.apache.spark.shuffle.sort.BypassMergeSortShuffleWriter.write(BypassMergeSortShuffleWriter.java:140)
  at org.apache.spark.shuffle.ShuffleWriteProcessor.write(ShuffleWriteProcessor.scala:59)
  at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:99)
  at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:52)
  at org.apache.spark.scheduler.Task.run(Task.scala:131)
  at org.apache.spark.executor.Executor$TaskRunner.$anonfun$run$3(Executor.scala:506)
  at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1462)
  at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:509)
  at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
  at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
  at java.lang.Thread.run(Thread.java:750)
Caused by: java.lang.ArrayIndexOutOfBoundsException: 0
  at ai.rapids.cudf.Table.<init>(Table.java:96)
  at ai.rapids.cudf.Table.readORC(Table.java:1289)
  at com.nvidia.spark.rapids.OrcCommonFunctions.$anonfun$decodeToTable$2(GpuOrcScan.scala:843)
  at com.nvidia.spark.rapids.Arm$.withResource(Arm.scala:29)
  at com.nvidia.spark.rapids.OrcCommonFunctions.$anonfun$decodeToTable$1(GpuOrcScan.scala:842)
  at com.nvidia.spark.rapids.RmmRapidsRetryIterator$NoInputSpliterator.next(RmmRapidsRetryIterator.scala:363)
  at com.nvidia.spark.rapids.RmmRapidsRetryIterator$RmmRapidsRetryIterator.next(RmmRapidsRetryIterator.scala:528)
  at com.nvidia.spark.rapids.RmmRapidsRetryIterator$RmmRapidsRetryAutoCloseableIterator.next(RmmRapidsRetryIterator.scala:458)
  at com.nvidia.spark.rapids.RmmRapidsRetryIterator$.drainSingleWithVerification(RmmRapidsRetryIterator.scala:275)
  at com.nvidia.spark.rapids.RmmRapidsRetryIterator$.withRetryNoSplit(RmmRapidsRetryIterator.scala:181)
  at com.nvidia.spark.rapids.OrcCommonFunctions.decodeToTable(GpuOrcScan.scala:840)
  ... 36 more
```

**Steps/Code to reproduce bug**
This can be reproduced with the tests/src/test/resources/schena-cant-prune.orc file in a Spark Shell session by executing something like this:
```
spark.read.orc("tests/src/test/resources/schema-cant-prune.orc").count
```

--- a/integration_tests/src/main/python/orc_test.py
+++ b/integration_tests/src/main/python/orc_test.py
@@ -424,6 +424,14 @@ def test_missing_column_names(spark_tmp_table_factory, reader_confs):
         reader_confs)
 
 @pytest.mark.parametrize('reader_confs', reader_opt_confs, ids=idfn)
+def test_missing_column_names_count(spark_tmp_table_factory, reader_confs):
+    table_name = spark_tmp_table_factory.get()
+    with_cpu_session(lambda spark : setup_orc_file_no_column_names(spark, table_name))
+    assert_gpu_and_cpu_row_counts_equal(
+        lambda spark : spark.sql("SELECT * FROM {}".format(table_name)),
+        reader_confs)
+
+@pytest.mark.parametrize('reader_confs', reader_opt_confs, ids=idfn)
 def test_missing_column_names_with_schema(spark_tmp_table_factory, spark_tmp_path, reader_confs):
     table_name = spark_tmp_table_factory.get()
     table_location = spark_tmp_path + "/ORC_DATA"
@@ -432,6 +440,14 @@ def test_missing_column_names_with_schema(spark_tmp_table_factory, spark_tmp_pat
         lambda spark : spark.read.schema("a int, b string, c int").orc(table_location),
         reader_confs)
 
+@pytest.mark.parametrize('reader_confs', reader_opt_confs, ids=idfn)
+def test_missing_column_names_count_with_schema(spark_tmp_table_factory, spark_tmp_path, reader_confs):
+    table_name = spark_tmp_table_factory.get()
+    table_location = spark_tmp_path + "/ORC_DATA"
+    with_cpu_session(lambda spark : setup_orc_file_no_column_names(spark, table_name, table_location))
+    assert_gpu_and_cpu_row_counts_equal(
+        lambda spark : spark.read.schema("a int, b string, c int").orc(table_location),
+        reader_confs)
 
 def setup_orc_file_with_column_names(spark, table_name):
     drop_query = "DROP TABLE IF EXISTS {}".format(table_name)

--- a/sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOrcScan.scala
+++ b/sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOrcScan.scala
@@ -1188,7 +1188,10 @@ private case class GpuOrcFileFilterHandler(
         // reader to EmptyPartitionReader for throwing exception
         null
       } else {
-        val (requestedColIds, canPruneCols) = resultedColPruneInfo.get
+        val requestedColIds = resultedColPruneInfo.get._1
+        // Normally without column names we cannot prune the file schema to the read schema,
+        // but if no columns are requested from the file (e.g.: row count) then we can prune.
+        val canPruneCols = resultedColPruneInfo.get._2 || requestedColIds.isEmpty
         OrcUtils.orcResultSchemaString(canPruneCols, dataSchema, readDataSchema,
           partitionSchema, conf)
         assert(requestedColIds.length == readDataSchema.length,

